File: test/language/structure_types/python/StructureInnerClassesClashingTest.py
import zserio

import StructureTypes

class StructureInnerClassesClashingTest(StructureTypes.TestCase):
    def testWriteReadArrayType(self):
        testStructure = self.api.ArrayType_array([1, 2, 3, 4])

        writer = zserio.BitStreamWriter()
        testStructure.write(writer)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readTestStructure = self.api.ArrayType_array.from_reader(reader)
        self.assertEqual(testStructure, readTestStructure)

    def testWriteReadOffsetChecker(self):
        testStructure = self.api.OffsetChecker_array([0] * 4, [1, 2, 3, 4])

        writer = zserio.BitStreamWriter()
        testStructure.initialize_offsets(writer.bitposition)
        testStructure.write(writer)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readTestStructure = self.api.OffsetChecker_array.from_reader(reader)
        self.assertEqual(testStructure, readTestStructure)

    def testWriteReadOffsetInitializer(self):
        testStructure = self.api.OffsetInitializer_array([0] * 4, [1, 2, 3, 4])

        writer = zserio.BitStreamWriter()
        testStructure.initialize_offsets(writer.bitposition)
        testStructure.write(writer)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readTestStructure = self.api.OffsetInitializer_array.from_reader(reader)
        self.assertEqual(testStructure, readTestStructure)

File: test/language/structure_types/cpp/StructureInnerClassesClashingTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "structure_types/structure_inner_classes_clashing/ArrayType_array.h"
#include "structure_types/structure_inner_classes_clashing/OffsetChecker_array.h"
#include "structure_types/structure_inner_classes_clashing/OffsetInitializer_array.h"

namespace structure_types
{
namespace structure_inner_classes_clashing
{

using allocator_type = ArrayType_array::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class StructureInnerClassesClashingTest : public ::testing::Test
{
protected:
    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

TEST_F(StructureInnerClassesClashingTest, writeReadArrayType)
{
    ArrayType_array testStructure{vector_type<uint32_t>{1, 2, 3, 4}};

    zserio::BitStreamWriter writer(bitBuffer);
    testStructure.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    ArrayType_array readTestStructure(reader);
    ASSERT_EQ(testStructure, readTestStructure);
}

TEST_F(StructureInnerClassesClashingTest, writeReadOffsetChecker)
{
    OffsetChecker_array testStructure;
    testStructure.setArray(vector_type<uint32_t>{1, 2, 3, 4});
    testStructure.getOffsets().resize(testStructure.getArray().size());
    testStructure.initializeOffsets();

    zserio::BitStreamWriter writer(bitBuffer);
    testStructure.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    OffsetChecker_array readTestStructure(reader);
    ASSERT_EQ(testStructure, readTestStructure);
}

TEST_F(StructureInnerClassesClashingTest, writeReadOffsetInitializer)
{
    OffsetInitializer_array testStructure;
    testStructure.setArray(vector_type<uint32_t>{1, 2, 3, 4});
    testStructure.getOffsets().resize(testStructure.getArray().size());
    testStructure.initializeOffsets();

    zserio::BitStreamWriter writer(bitBuffer);
    testStructure.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    OffsetInitializer_array readTestStructure(reader);
    ASSERT_EQ(testStructure, readTestStructure);
}

} // namespace structure_inner_classes_clashing
} // namespace structure_types
